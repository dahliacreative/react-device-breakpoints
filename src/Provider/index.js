import React, { useState, useEffect } from 'react'
import { Provider } from '../Context'

const createListener = (onChange, bp) => ({ matches }) => {
    onChange({
        [bp.name]: matches
    })
}

const Breakpoints = ({children, ...props}) => {

    const breakpoints = Object.entries(props).map((key, query) => ({
        key,
        query: window.matchMedia(query)
    }))

    const defaultState = breakpoints.reduce((acc, bp) => ({
        ...acc,
        [bp.key]: bp.query.matches
    }), {
        isTouchDevice: /Mobi|Tablet|iPad|iPhone|Android/.test(window.navigator.userAgent)
    })

    const [device, updateDevice] = useState(defaultState)

    const onChange = bp => {
        updateDevice({
            ...device,
            ...bp
        })
    }

    useEffect(() => {
        const listeners = breakpoints.map((bp, i) => {
            const listener = createListener(onChange, bp)
            bp.query.addListener(listener)
            return listener
        })

        return () => {
            breakpoints.forEach((bp, i) => {
                bp.query.removeListener(listeners[i])
            })
        }
    }, [])

    return <Provider value={device}>{children}</Provider>
}

export default Breakpoints