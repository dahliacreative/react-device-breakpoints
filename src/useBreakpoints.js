import { useContext } from 'react'
import Context from './context'

const useBreakpoints = () => useContext(Context)

export default useBreakpoints