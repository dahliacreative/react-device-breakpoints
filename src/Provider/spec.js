import React from 'react'
import { mount } from 'enzyme'

import Provider from './'
import useBreakpoints from '../Hooks'

const breakpoints = {
    isDesktop: '(min-width: 1024px)',
    isTablet: '(max-width: 1023px) and (min-width: 768px)',
    isMobile: '(max-width: 767px)'
}

const Consumer = () => {
    const device = useBreakpoints()
    return <p>{JSON.stringify(device)}</p>
}

describe('<Provider />', () => {
    it('renders given children', () => {
        const component = mount(
            <Provider>
                <h1>React Device Breakpoints FTW!</h1>
            </Provider>
        ) 
        expect(component).toMatchSnapshot()
    })

    it('renders provider with given queries', () => {
        const component = mount(
            <Provider {...breakpoints}>
                <h1>React Device Breakpoints FTW!</h1>
                <Consumer />
            </Provider>
        )
        const consumer = component.find(Consumer)
        expect(component).toMatchSnapshot()
        expect(consumer).toMatchSnapshot()
    })

    it('removes listeners when unmounting', () => {
        const component = mount(
            <Provider {...breakpoints}>
                <h1>React Device Breakpoints FTW!</h1>
            </Provider>
        )
        component.unmount()
        expect(component).toMatchSnapshot()
    })
})