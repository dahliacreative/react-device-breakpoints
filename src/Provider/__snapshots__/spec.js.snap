// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Provider /> removes listeners when unmounting 1`] = `ReactWrapper {}`;

exports[`<Provider /> renders given children 1`] = `ReactWrapper {}`;

exports[`<Provider /> renders provider with given queries 1`] = `ReactWrapper {}`;

exports[`<Provider /> renders provider with given queries 2`] = `ReactWrapper {}`;
